### PR TITLE
Simplify memory management in example 1

### PR DESCRIPTION
**Pull request by flomnes**
Memory allocation on the heap is not necessary and can be confusing for new users.

 | PR | Author | Editor | Reviewers | Assignment | Approval | Merge| 
 | --- | --- | --- | --- | --- | --- | --- | 
| [#1587](https://github.com/mfem/mfem/pull/1587) | @flomnes | @tzanio | @jakubcerveny + @YohannDudouit | 07/07/20 | 07/21/20 | 07/27/20 | |

### CODE REVIEW

**Comment by tzanio** [comment]
Please add matching changes in `ex1p.cpp`

**Comment by flomnes** [comment]
@rw-anderson Probably the best way would be to pass object by (possibly const) reference, which is more natural in C++. I understand though it would break compatibility with existing code.

File: examples/ex1.cpp [hunk 1]
```diff
@@ -102,88 +102,88 @@ int main(int argc, char *argv[])
...
+   if (static_cond) { a.EnableStaticCondensation(); }
+   a.Assemble();
 
    OperatorPtr A;
    Vector B, X;
-   a->FormLinearSystem(ess_tdof_list, x, *b, A, X, B);
+   a.FormLinearSystem(ess_tdof_list, x, b, A, X, B);
 
-   cout << "Size of linear system: " << A->Height() << endl;
+   cout << "Size of linear system: " << a.Height() << endl;
```
Comment (v-dobrev) [inline examples/ex1.cpp:186]: This change is not accurate: if static condensation is enabled `A->Height()` is not the same as `a.Height()`.

File: examples/ex1.cpp [hunk 2]
```diff
@@ -102,88 +102,88 @@ int main(int argc, char *argv[])
...
-         mesh->UniformRefinement();
+         mesh.UniformRefinement();
       }
    }
 
    // 5. Define a finite element space on the mesh. Here we use continuous
    //    Lagrange finite elements of the specified order. If order < 1, we
    //    instead use an isoparametric/isogeometric space.
-   FiniteElementCollection *fec;
+   FiniteElementCollection *fec = nullptr;
```
Comment (v-dobrev) [inline examples/ex1.cpp:124]: This pointer is initialized in all cases in the following `if` statement, so this initialization is redundant.

File: examples/ex1.cpp [hunk 3]
```diff
@@ -232,15 +232,11 @@ int main(int argc, char *argv[])
...
+      sol_sock << "solution\n" << mesh << x << flush;
    }
 
    // 15. Free the used memory.
-   delete a;
-   delete b;
-   delete fespace;
-   if (order > 0) { delete fec; }
-   delete mesh;
+   delete fec;
```
Comment (v-dobrev) [inline examples/ex1.cpp:239]: The condition for deleting `fec` should not be changed.

File: examples/ex1p.cpp [hunk 4]
```diff
@@ -111,46 +111,45 @@ int main(int argc, char *argv[])
...
+         mesh.UniformRefinement();
       }
    }
 
    // 6. Define a parallel mesh by a partitioning of the serial mesh. Refine
    //    this mesh further in parallel to increase the resolution. Once the
    //    parallel mesh is defined, the serial mesh can be deleted.
-   ParMesh *pmesh = new ParMesh(MPI_COMM_WORLD, *mesh);
-   delete mesh;
+   ParMesh pmesh(MPI_COMM_WORLD, mesh);
```
Comment (v-dobrev) [inline examples/ex1p.cpp:133]: Use `mesh.Clear()` to free the dynamic memory used by `mesh`.

File: examples/ex1p.cpp [hunk 5]
```diff
@@ -111,46 +111,45 @@ int main(int argc, char *argv[])
...
-         pmesh->UniformRefinement();
+         pmesh.UniformRefinement();
       }
    }
 
    // 7. Define a parallel finite element space on the parallel mesh. Here we
    //    use continuous Lagrange finite elements of the specified order. If
    //    order < 1, we instead use an isoparametric/isogeometric space.
-   FiniteElementCollection *fec;
+   FiniteElementCollection* fec;
```
Comment (v-dobrev) [inline examples/ex1p.cpp:145]: Revert this -- this is the style used by mfem.

File: examples/ex1p.cpp [hunk 6]
```diff
@@ -263,16 +262,10 @@ int main(int argc, char *argv[])
...
-      sol_sock << "solution\n" << *pmesh << x << flush;
+      sol_sock << "solution\n" << pmesh << x << flush;
    }
 
    // 17. Free the used memory.
-   delete a;
-   delete b;
-   delete fespace;
-   if (order > 0) { delete fec; }
-   delete pmesh;
```
Comment (v-dobrev) [inline examples/ex1p.cpp:274]: The pointer `fec` still needs to be deleted as before.

File: examples/ex1.cpp (hunk 3 above)
Comment (YohannDudouit) [reply examples/ex1.cpp:239]: The previous condition was ignoring the `new` line 136 though.

**Comment by flomnes** [comment]
@v-dobrev I took your remarks into account. fec is now deleted only when it was dynamically allocated.

**Comment by YohannDudouit** [comment]
@flomnes Can you fix the style issues?

**Comment by flomnes** [comment]
> @flomnes Can you fix the style issues?

Done

**Comment by tzanio** [comment]
LGTM

**Comment by tzanio** [comment]
Merged in `next` for testing...

**Comment by tzanio** [comment]
Thanks for this contribution @flomnes!

**Comment by benzwick** [comment]
Which of these two styles is preferred in MFEM? The examples are a bit inconsistent, e.g.

- ex1: `Mesh mesh(mesh_file, 1, 1);`
- ex2: `Mesh *mesh = new Mesh(mesh_file, 1, 1);`
- ex17: `Mesh mesh(mesh_file, 1, 1);`

The reason I stumbled upon this issue is because I was trying to add some functionality from ex17 (DG elasticity) to ex2 (CG elasticity) which use different styles. What are the pros and cons of each approach aside from not having to delete the objects at the end? Are there any examples where heap allocation would be preferred?

**Comment by tzanio** [comment]
There is really no preference, it is up to you and what you are comfortable with. If you manage memory yourself, you have to be careful about leaks and data ownership.

Most of the code uses pointers and explicit allocation and deallocation. In a few of the examples however, it was deemed (see e.g. this PR) simpler to avoid it.

**Comment by benzwick** [comment]
Thanks for the clarification. Would it be worth adding this to the documentation somewhere?

**Comment by tzanio** [comment]
If it will help users, yes. Not sure what's the best place though.

**Pull request merged**